ddclient.in

```diff
@@ -5912,6 +5923,81 @@ sub nic_googledomains_update {
+## nic_henet_examples
+##
+## written by Indrajit Raychaudhuri
+##
+######################################################################
+sub nic_henet_examples {
+    return <<"EoEXAMPLE";
+o 'henet'
+
+The 'henet' protocol is used by DNS service offered by dns.he.net.
+
+Configuration variables applicable to the 'henet' protocol are:
+  protocol=henet               ##
+  password=service-password    ## the password provided by the admin interface
+  fully.qualified.host         ## the host registered with the service.
+
+Example ${program}.conf file entries:
+  ## single host update
+  protocol=henet,                                              \\
+  password=my-genereated-password                              \\
+  myhost.example.com
+EoEXAMPLE
+}
+######################################################################
+## nic_henet_update
+######################################################################
+sub nic_henet_update {
+    debug("\nnic_henet_update -------------------");
+
+    my %errors = (
+       'badauth' => 'Bad authorization (username or password)',
+       'badsys' => 'The system parameter given was not valid',
+       'nohost' => 'The hostname specified does not exist in the database',
+       'abuse' => 'The hostname specified is blocked for abuse',
+       'nochg' => 'No update required; unnecessary attempts to change to the current address are considered abusive',
+    );
+
+    for my $h (@_) {
+        # The IPv4 and IPv6 addresses must be updated in separate API call.
+        for my $ipv ('4', '6') {
+            my $ip = delete($config{$h}{"wantipv$ipv"}) or next;
+            info("Setting IPv%s address to %s for %s", $ipv, $ip, $h);
+            verbose("UPDATE:", "updating %s", $h);
+            my $reply = geturl(
+                proxy => opt('proxy'),
+                url => "https://$config{$h}{'server'}/nic/update?hostname=$h&myip=$ip",
+                login => $h,
+                password => $config{$h}{'password'},
+            ) // '';
+            if ($reply eq '') {
+                failed("updating %s: Could not connect to %s.", $h, $config{$h}{'server'});
+                next;
+            }
+            next if !header_ok($h, $reply);
+            (my $body = $reply) =~ s/.*?\n\n//s;
+            my ($line) = split(/\n/, $body, 2);
+            my ($status, $returnedip) = split(/ /, lc($line));
+            $status = 'good' if $status eq 'nochg';
+            $config{$h}{"status-ipv$ipv"} = $status;
+            if ($status ne 'good') {
+                if (exists($errors{$status})) {
+                    failed("updating %s: %s: %s", $h, $status, $errors{$status});
+                } else {
+                    failed("updating %s: unexpected status: %s", $h, $line);
+                }
+                next;
+            }
+            success("updating %s: %s: IPv%s address set to %s", $h, $status, $ipv, $returnedip);
```

Comment:
The he.net documentation doesn't specify what the server returns in the HTTP body, so I wonder if it would be better to only pay attention to the HTTP status code:

```suggestion
            success("updating %s: %s: IPv%s address set to %s", $h, $status, $ipv, $returnedip);
            $config{$h}{"status-ipv$ipv"} = 'good';
```

Reply:
While he.net documentation is very sparse, it _does_ respond with actual status in HTTP response body like so:

1. Status `good`:
```
RECEIVE:   HTTP/1.1 200 OK
RECEIVE: | Date: Wed, 26 Jun 2024 00:44:09 GMT
RECEIVE: | Server: dns.he.net v0.0.1
RECEIVE: | Email: DNS Administrator <dnsadmin@he.net>
RECEIVE: | Cache-Control: no-cache, must-revalidate
RECEIVE: | Expires: Thu, 26 Jun 2025 00:44:09 GMT
RECEIVE: | Content-Length: 43
RECEIVE: | Content-Type: text/html
RECEIVE: | 
RECEIVE: | good 2001:123:aaaa:aaaa:aaaa:aaaa:aaaa:aaaa
```

2. Status `nochg`:
```
RECEIVE:   HTTP/1.1 200 OK
RECEIVE: | Date: Wed, 26 Jun 2024 00:44:09 GMT
RECEIVE: | Server: dns.he.net v0.0.1
RECEIVE: | Email: DNS Administrator <dnsadmin@he.net>
RECEIVE: | Cache-Control: no-cache, must-revalidate
RECEIVE: | Expires: Thu, 26 Jun 2025 00:44:09 GMT
RECEIVE: | Content-Length: 19
RECEIVE: | Content-Type: text/html
RECEIVE: | 
RECEIVE: | nochg 123.123.123.123
```

3. Status `badauth`:
```
RECEIVE:   HTTP/1.1 200 OK
RECEIVE: | Date: Wed, 26 Jun 2024 00:51:06 GMT
RECEIVE: | Server: dns.he.net v0.0.1
RECEIVE: | Email: DNS Administrator <dnsadmin@he.net>
RECEIVE: | Cache-Control: no-cache, must-revalidate
RECEIVE: | Expires: Thu, 26 Jun 2025 00:51:06 GMT
RECEIVE: | Content-Length: 7
RECEIVE: | Content-Type: text/html
RECEIVE: | 
RECEIVE: | badauth
```

4. Status `abuse`: (this is an old message I pull up from log)
```
RECEIVE:   HTTP/1.1 200 OK
RECEIVE: | Date: Wed, 17 Apr 2024 04:23:16 GMT
RECEIVE: | Server: dns.he.net v0.0.1
RECEIVE: | Email: DNS Administrator <dnsadmin@he.net>
RECEIVE: | Cache-Control: no-cache, must-revalidate
RECEIVE: | Expires: Thu, 18 Jun 2025 04:23:16 GMT
RECEIVE: | Content-Length: 7
RECEIVE: | Content-Type: text/html
RECEIVE: | 
RECEIVE: | abuse
```

However, the HTTP response header is always `200 OK`, so that's not useful.

Reply:
Thanks for investigating.  Maybe add a comment that says that it replies with 200 OK in certain error conditions?

Reply:
I pushed a fixup commit that adds a comment explaining why the body must be checked.